import "google/protobuf/empty.proto";

package NKikimrChangeExchange;
option java_package = "ru.yandex.kikimr.proto";

message TDataChange {
    message TSerializedCells {
        // Column tags
        repeated uint32 Tags = 1; 
        // Serialized values (TSerializedCellVec)
        optional bytes Data = 2;
    }

    /* key columns */
    optional TSerializedCells Key = 1;

    /* value columns (doesn't include key columns) */
    // old image
    optional TSerializedCells OldImage = 2;
    // new image
    optional TSerializedCells NewImage = 3;
    // Row operation
    // Cells (if any) can be empty if:
    // - there is no change in data-columns of AsyncIndex;
    // - there is no need to save update to Cdc Stream (e.g. in "keys only" mode).
    oneof RowOperation {
        // apply specified values
        TSerializedCells Upsert = 4;
        // remove entire row
        google.protobuf.Empty Erase = 5;
        // reset all columns & apply specified values
        TSerializedCells Reset = 6;
    }
}

message TCdcHeartbeat {
}

message TChangeRecord {
    optional uint64 Order = 1;
    optional uint64 Group = 2;
    optional uint64 Step = 3;
    optional uint64 TxId = 4;
    optional uint64 PathOwnerId = 5;
    optional uint64 LocalPathId = 6;
    oneof Kind {
        TDataChange AsyncIndex = 7;
        TDataChange CdcDataChange = 8;
        TCdcHeartbeat CdcHeartbeat = 9;
        TDataChange IncrementalRestore = 10;
    }
}

message TEvHandshake {
    // tablet id
    optional uint64 Origin = 1;
    // tablet generation
    optional uint64 Generation = 2;
}

message TEvApplyRecords {
    // tablet id
    optional uint64 Origin = 1;
    // tablet generation
    optional uint64 Generation = 2;
    // records
    repeated TChangeRecord Records = 3;
}

message TEvStatus {
    enum EStatus {
        STATUS_UNKNOWN = 0;
        STATUS_OK = 1;
        STATUS_REJECT = 2;
    }

    enum EReason {
        REASON_NONE = 0;
        REASON_UNKNOWN_ORIGIN = 1;
        REASON_STALE_ORIGIN = 2;
        REASON_ORDER_VIOLATION = 3;
        REASON_UNEXPECTED_KIND = 4;
        REASON_UNEXPECTED_ROW_OPERATION = 5;
        REASON_SCHEME_ERROR = 6;
        REASON_WRONG_STATE = 7;
    }

    message TRecordStatus {
        optional uint64 Order = 1;
        optional EStatus Status = 2;
        optional EReason Reason = 3;
    }

    optional EStatus Status = 1;
    optional EReason Reason = 2;
    repeated TRecordStatus RecordStatuses = 3;
    optional uint64 LastRecordOrder = 4;
}

message TEvActivateSender {
    optional uint64 Origin = 1;
}

message TEvActivateSenderAck {
    optional uint64 Origin = 1;
}
